### PR TITLE
Fix issues around location in nearby activity

### LINKED ISSUE #982: Optional Permissions in Nearby Places activity?
On opening `NearbyActivity` when location is disabled, the alert dialog shows the following permission. The text is somehow misleading as its not possible to show nearby places without location permissions in `NearbyActivity`. 

optional_permissions

IMO it makes more sense to ask for location permissions, mandatorily in `NearbyActivity`. 
- If the user grants location permissions, load nearby places
- Else finish the activity. 

Comments, @nicolas-raoul @misaochan

### PR DESCRIPTION
**Pull request by maskaravivek**
- Fixes the issue where nearby list was not loaded in the first attempt. 

How to reproduce?
- Initially phone's GPS is off or the Commons app doesn't have location permissions.
- Either of these things are provided at run time.
- The nearby list just keeps loading. 
- It loads after killing the app. 

The above issue is now fixed. 

- App was continuously listening for location updates and draining a lot of battery

How do you know?
- The location Setting page in the your phone tags Commons app as high battery usage application. 

The above should be hopefully fixed. 
- The app now stops listening for location updates as soon as the nearby list is populated. 
- It again registers the location manager when the refresh button is clicked. 

- App now uses multiple providers to fetch the location. 

Why?

`GPS_PROVIDER` is usually slower. Now the app would be using both `NETWORK_PROVIDER` for coarse location and `GPS_PROVIDER` for finer location. Upon receiving a location update, it will determine if the newer location is better than the last location. If it is, the newer location would be used. 

Also fixes #982.

### CODE REVIEW

**Comment by maskaravivek** [comment]
Will be fixing #978 after this is merged.

File: app/src/androidTest/java/fr/free/nrw/commons/NearbyActivityTest.java [hunk 1]
```diff
@@ -24,7 +31,34 @@
...
     public void testActivityLaunch() {
+        onView(withText(R.string.gps_disabled))
+                .check(matches(isDisplayed()));
+        onView(withId(android.R.id.button2)).perform(click());
         onView(withText(R.string.title_activity_nearby))
-                .check(ViewAssertions.matches(isDisplayed()));
+                .check(matches(isDisplayed()));
     }
+
+//    private void turnGPSOn(Context context){
```
Comment (misaochan) [inline app/src/androidTest/java/fr/free/nrw/commons/NearbyActivityTest.java]: Could I ask why this is commented out?

**Comment by misaochan** [comment]
Thanks for submitting this, @maskaravivek ! I'm hoping to test it this weekend if I can, and then maybe we could include it in our release (planned for early next week). Some conflicts need to be resolved though, after the recent test PR merge.

**Comment by misaochan** [comment]
Tested this on real device running Android 7.0 and works well for me. Happy to merge once conflicts are resolved. :)

File: app/src/androidTest/java/fr/free/nrw/commons/NearbyActivityTest.java (hunk 1 above)
Comment (maskaravivek) [reply app/src/androidTest/java/fr/free/nrw/commons/NearbyActivityTest.java]: I was trying out something. It doesn't matter now. We have @psh's test to replace ours.

**Comment by maskaravivek** [comment]
@misaochan I have rebased the branch.

**Pull request merged**